File: 200-number-of-islands/main.go
func numIslands(grid [][]byte) int {
    
    rows, cols := len(grid), len(grid[0])

    if rows == 0 {
        return 0
    }

    visited := make(map[[2]int]bool)
    islands := 0

    var dfs func(row, col int)
    dfs = func (row, col int) {
        if visited[[2]int{row, col}] || row >= rows || row < 0 || col >= cols || col < 0 || grid[row][col] == '0' {
            return
        }

        if _, seen := visited[[2]int{row, col}]; seen {
            return
        }

        visited[[2]int{row, col}] = true
        directions := [][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
        for _, dir := range directions {
            r, c := row+dir[0], col+dir[1]
            dfs(r, c)
        }
    }

    for row := range(rows) {
        for col := range(cols) {
            if grid[row][col] == '1' && !visited[[2]int{row, col}] {
                dfs(row, col)
                islands++
            }
        }
    }

    return islands

}

File: 200-number-of-islands/main.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        if not grid:
            return 0

        rows, cols = len(grid), len(grid[0])
        visited = set()
        islands = 0

        def dfs(row: int, col: int) -> None:

            if (row, col) in visited or row >= rows or col >= cols or col < 0 or row < 0 or grid[row][col] == "0":
                return
            
            visited.add((row, col))

            directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            for (x, y) in directions:
                dfs(row + x, col + y)

        for row in range(rows):
            for col in range(cols):
                if grid[row][col] == "1" and (row, col) not in visited:
                    dfs(row, col)
                    islands += 1
        

        return islands

